eng/scripts/Invoke-GenerateAndBuildV2.ps1

```diff
@@ -108,6 +109,39 @@ if ($inputFileToGen) {
     UpdateExistingSDKByInputFiles -inputFilePaths $inputFileToGen -sdkRootPath $sdkPath -headSha $commitid -repoHttpsUrl $repoHttpsUrl -downloadUrlPrefix "$downloadUrlPrefix" -generatedSDKPackages $generatedSDKPackages
 }
 
+# generate sdk from cadl file
+if ($relatedCadlProjectFolder) {
+    for ($i = 0; $i -le $relatedCadlProjectFolder.Count - 1; $i++) {
+        $caldFolder = (Join-Path $swaggerDir $relatedCadlProjectFolder[$i]) -replace "\\", "/"
+        $newPackageOutput = "newPackageOutput.json"
+
+        Push-Location $caldFolder
+        $cadlProjectYaml = Get-Content -Path "$caldFolder/cadl-project.yaml" -Raw
+
+        Install-ModuleIfNotInstalled "powershell-yaml" "0.4.1" | Import-Module
+        $yml = ConvertFrom-YAML $cadlProjectYaml
+        $sdkFolder = $yml["emitters"]["@azure-tools/cadl-csharp"]["sdk-folder"]
+        $projectFolder = (Join-Path $sdkPath $sdkFolder)
+        $projectFolder = $projectFolder -replace "\\", "/"
+        if ($projectFolder) {
+            $directories = $projectFolder.Split("/");
+            $count = $directories.Count
+            $projectFolder = $directories[0 .. ($count-2)] -join "/"
+            $service = $directories[-3];
+            $namespace = $directories[-2];
+        }
+        New-CADLPackageFolder -service $service -namespace $namespace -sdkPath $sdkPath -cadlInput $caldFolder/main.cadl -outputJsonFile $newpackageoutput
+        $newPackageOutputJson = Get-Content $newPackageOutput | Out-String | ConvertFrom-Json
+        $relativeSdkPath = $newPackageOutputJson.path
+        npm install
+        npx cadl compile --output-path $sdkPath --emit @azure-tools/cadl-csharp .
+        if ( !$?) {
+            Throw "Failed to generate sdk for cadl. exit code: $?"
+        }
+        GeneratePackage -projectFolder $projectFolder -sdkRootPath $sdkPath -path $relativeSdkPath -downloadUrlPrefix "$downloadUrlPrefix" -skipGenerate -generatedSDKPackages $generatedSDKPackages
+        Pop-Location
```

Comment:
If anything fails above and `$ErrorActionPreference` is `Stop`, you'll not pop the location. I recommend against this pattern. Use the full directories, and if tools can't handle that they need to be fixed. For example, I can run `msbuild` or `dotnet build` on a .csproj in any directory from any other directory and it works. Tools should do this. If `npm` is the reason above, it takes full paths and output directories you can set. If that's not otherwise reasonable, then on line 118 you should do something like:
```powershell
push-location $cadlFolder
trap { pop-location }
```
This will make sure that any error pops the location regardless of where it happens.

Reply:
Thanks @heaths  we need to install all the modules defined in `package.json`, but `npm install` cannot specify the file, so we need to enter the directory which contains the `package.json`. 
I will use `trap` to make sure `pop-location` will be executed when script is exit in unnormal way.